# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/workers/ScriptLoader.cpp
# Commit: 997e9ca2e1e9
# Full Hash: 997e9ca2e1e9d636f8f956151db81ca2a8a60c82
# Author: Yulia Startsev <ystartsev@mozilla.com>
# Date: 2023-01-19 04:46:07
# Regressor Bug: 1247687
# File Overlap Count: 1
# Description:
#   Bug 1247687 - Implement InitModuleLoader methods for WorkerScriptLoader and WorkerGlobalScope; r=jonco
#   
#   This implements a method to initialize the moduleLoader for workers. This will initialize only once, for all worker types (module and classic).
#   
#   Depends on D147324
# ==============================================================================

diff -r ebeb65b42212 -r 997e9ca2e1e9 dom/workers/ScriptLoader.cpp
--- a/dom/workers/ScriptLoader.cpp	Wed Jan 18 13:46:29 2023 +0000
+++ b/dom/workers/ScriptLoader.cpp	Wed Jan 18 13:46:29 2023 +0000
@@ -411,8 +411,24 @@
   }
 
   nsIGlobalObject* global = GetGlobal();
+  mController = global->GetController();
+  // Set up the module loader, if it has not been yet.
+  // TODO: Implement this for classic scripts when dynamic imports are added.
+  if (aWorkerPrivate->WorkerType() == WorkerType::Module) {
+    InitModuleLoader();
+  }
+}
 
-  mController = global->GetController();
+void WorkerScriptLoader::InitModuleLoader() {
+  mWorkerRef->Private()->AssertIsOnWorkerThread();
+  RefPtr<WorkerModuleLoader> moduleLoader =
+      new WorkerModuleLoader(this, GetGlobal(), mSyncLoopTarget.get());
+
+  if (mWorkerScriptType == WorkerScript) {
+    mWorkerRef->Private()->GlobalScope()->InitModuleLoader(moduleLoader);
+    return;
+  }
+  mWorkerRef->Private()->DebuggerGlobalScope()->InitModuleLoader(moduleLoader);
 }
 
 void WorkerScriptLoader::CreateScriptRequests(